pkg/controller/core/resourceflavor_controller.go

```diff
@@ -53,20 +56,37 @@ func (r *ResourceFlavorReconciler) Create(e event.CreateEvent) bool {
 	if !match {
 		return false
 	}
+
 	log := r.log.WithValues("resourceFlavor", klog.KObj(flv))
 	log.V(2).Info("ResourceFlavor create event")
-	r.cache.AddOrUpdateResourceFlavor(flv.DeepCopy())
+
+	// As long as one clusterQueue got ready, we should inform clusterQueue controller to broadcast the event.
+	// Ignore the second value for add/update flavor will not cause clusterQueue not ready.
+	cqNames, _ := r.cache.AddOrUpdateResourceFlavor(flv.DeepCopy())
+	r.qManager.UpdateClusterQueuesStatus(cqNames, true)
+	if len(cqNames) > 0 {
+		r.qManager.Broadcast()
+	}
 	return false
 }
 
+// TODO: When ResourceFlavors are used by ClusterQueues, they shouldn't been deleted directly.
 func (r *ResourceFlavorReconciler) Delete(e event.DeleteEvent) bool {
 	flv, match := e.Object.(*kueue.ResourceFlavor)
 	if !match {
 		return false
 	}
+
 	log := r.log.WithValues("resourceFlavor", klog.KObj(flv))
 	log.V(2).Info("ResourceFlavor delete event")
-	r.cache.DeleteResourceFlavor(flv)
+
+	// TODO: When finalizer implemented in resourceFlavor controller, we'll no longer need notReadyNames.
+	readyNames, notReadyNames := r.cache.DeleteResourceFlavor(flv)
+	r.qManager.UpdateClusterQueuesStatus(readyNames, true)
+	r.qManager.UpdateClusterQueuesStatus(notReadyNames, false)
```

Comment:
I think there is a potential for a race condition here.

The race could happen between updating the state in the cache and updating it in the manager; the fact that those two updates are not done while holding a common lock is going to cause race conditions  between the resourceFlavor controller and the clusterQueue controller.

I don't think we need to track the active state in the manager at all; in fact nothing should change in the queue manager; instead in the scheduler, [after we get the heads of the queues](https://github.com/kubernetes-sigs/kueue/blob/277a453a215f5935d266011e27a97c597cb2d7e7/pkg/scheduler/scheduler.go#L83), for each workload we check if its CQ exists in the snapshot or not.

Reply:
> I don't think we need to track the active state in the manager at all; in fact nothing should change in the queue manager; instead in the scheduler, [after we get the heads of the queues](https://github.com/kubernetes-sigs/kueue/blob/277a453a215f5935d266011e27a97c597cb2d7e7/pkg/scheduler/scheduler.go#L83), for each workload we check if its CQ exists in the snapshot or not.

This is my first design, but as you said, the potential data race is always exist, I think correctness here is more important than performance, and the loss is not very big.  I'll change as you advised.

Reply:
I think we can keep the check in the manager when getting the heads, this will be better especially for BestEffortFIFO. The check can be done from the cache by invoking `ClusterQueueActive`.

I feel that a dependency from the queue to the cache is ok (the queue is logically over the cache), but we could also pass in the function's reference to the manager rather than the whole cache to avoid that dependency. @alculquicondor do you think there is an issue with the queue to cache dependency?

Reply:
@kerthcet lets go with a function reference or interface instead of passing the whole cache.

Reply:
Or maybe follow this in the next PR. I'd like to implement this feature firstly.

Reply:
created issue here: https://github.com/kubernetes-sigs/kueue/issues/251